### DISCUSSION TITLE
error-stack: pattern for attaching to a bunch of errors?

### DISCUSSION THREAD

**Discussion by bjchambers** [hashintel/hash#1982] [error-stack (Help)]
A common pattern I see is having a method that attempts to read a file, and wanting to attach the file to any error within the method. I haven't found a good way to avoid duplicating things, and this often ends up with something like:

```
method1().attach_printable(path)?;
method2().attach_printable(path)?;
...
```

Sometimes there may be two or three printables that I wish to add. I find myself wishing I could register something as a printable to be attached if/when an error happens (lazily would be ideal, since often it may need to be cloned so the error can own it). Is there a pattern for this? Does anyone have ideas on how to avoid the repetition?

**Comment by TimDiekmann** [comment]
Hi, @bjchambers, and sorry for not responding for this long. I missed this question 😞 

Typically, if you find yourself adding the same attachments over and over again it typically is a sign that those functions should be moved into a separate function and add the attachments to that function call. Ideally, you want to attach _outside_ of the actual functions. For example, if you have a function to create a file at a given path `fn create_file(path: Path)`. You don't want to attach the `path` inside of `create_file` because the path might not be complete (additional context such as the CWD is missing here). Instead, you want to `create_file(path).attach_printable_lazy(|| ...)`. 

The above is not always possible so one solution would be to create a function that takes a `Report` so you can call `.map_err` on the function result. If you want to globally register such a function you would need to inject the context into that registry somehow so this would just move the problem to a different location.

Hope this will still help you, excited to hear what you have ended up with! 🙂
